### ISSUE TITLE
x/net context canceled

### ISSUE THREAD

**Issue by matthewceravolo** [golang/go#26356]
Please answer these questions before submitting your issue. Thanks!

### What version of Go are you using (`go version`)?
go version go1.10 linux/amd64

### Does this issue reproduce with the latest release?
Yes

### What operating system and processor architecture are you using (`go env`)?
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/matthew/.cache/go-build"
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/home/matthew/work"
GORACE=""
GOROOT="/usr/local/go"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build356255000=/tmp/go-build -gno-record-gcc-switches"

### What did you do?
used context.WithTimeout() to make requests to google calendar api and outlook calendar api

If possible, provide a recipe for reproducing the error.
A complete runnable program is good.
A link on play.golang.org is best.

### What did you expect to see?
Making requests using contexts with timeouts should cancel when the timeout is reached

### What did you see instead?
Contexts with timeouts are instantly failing with "context canceled" even though the timeout is set to time.Minute. The error goes away if I remove the timeout context and use one without any limit. It also _seems_ to be transient to some extent

**Comment by ianlancetaylor** [comment]
Sorry, but there isn't enough information here to say anything useful.  Please show us your code, or tell us how to recreate the problem.  Thanks.

**Comment by matthewceravolo** [comment]
import (
	"net/http"
	"os"
	"time"

	"golang.org/x/net/context"
	"golang.org/x/oauth2"
	"golang.org/x/oauth2/google"
)

var (
	backgroundContext = context.Background()
	oauthConfig       = &oauth2.Config{
		ClientID:     os.Getenv("GOOGLE_PROJECT_CLIENT_ID"),
		ClientSecret: os.Getenv("GOOGLE_PROJECT_SECRET_KEY"),
		Endpoint:     google.Endpoint,
	}
)

type calendarClient struct {
	*calendar.Service
	requestTimeout time.Duration
}

// NewCalendarClient creates a new authenticated Google Calendar client.
func NewCalendarClient(refreshToken string) (wf.Client, error) {
	client, err := newClient(refreshToken)
	if err != nil {
		return nil, err
	}

	service, err := calendar.New(client)
	if err != nil {
		return nil, err
	}

	return &calendarClient{Service: service, requestTimeout: time.Minute}, nil
}

func newClient(refreshToken string) (*http.Client, error) {
	contextWithTimeout, cancel := context.WithTimeout(backgroundContext, time.Minute)
	defer cancel()
	return oauthConfig.Client(contextWithTimeout, &oauth2.Token{RefreshToken: refreshToken}), nil
}

func (client *calendarClient) GetCalendars() (*calendar.CalendarList, error) {
	contextWithTimeout, cancel := context.WithTimeout(backgroundContext, time.Minute)
	defer cancel()
	return client.CalendarList.List().Context(contextWithTimeout).Do()
}

The issue is that when we call GetCalendars for example, it will instantly return "context canceled". When I use an empty context with no timeout, everything works as expected. This all began failing in the past couple of days since updates were made in the /x/net package

**Comment by meirf** [comment]
[`NewClient`](https://godoc.org/golang.org/x/oauth2#NewClient):

> The returned client is not valid beyond the lifetime of the context.

But your code has:
```
func newClient(refreshToken string) (*http.Client, error) {
    contextWithTimeout, cancel := context.WithTimeout(backgroundContext, time.Minute)
    defer cancel()
    return oauthConfig.Client(contextWithTimeout, &oauth2.Token{RefreshToken: refreshToken}), nil
}
```
The `defer cancel()` causes the context to be cancelled before the client even gets to `calendar.New`.

**Comment by bcmills** [comment]
I've seen this sort of issue crop up several times now.

I wonder if `context.Context` should record a bit of caller information (say, the PC of the immediate calling function) in `context.WithTimeout` and in calls to the `CancelFunc` returned by `context.WithCancel`. Then we could add a debugging hook to interrogate *why* a particular `context.Context` was cancelled.

Perhaps something along the lines of:
```go
package context

// A DoneReasoner describes the reason a Context is done.
// The Context implementations returned by this package implement DoneReasoner.
type DoneReasoner interface {
	DoneReason() string
}

// DoneReason returns a human-readable description of the reason that ctx.Done() is closed,
// or the empty string if ctx.Done() is not closed.
func DoneReason(ctx context.Context) string {
	select {
	case <-ctx.Done():
	default:
		return ""
	}
	if r, ok := ctx.(DoneReasoner); ok {
			return r.DoneReason()
	}
	return ctx.Err().Error()
}
```

**Comment by bcmills** [comment]
(CC @Sajmani @bradfitz for Context debuggability.)

**Comment by matthewceravolo** [comment]
@meirf I think there was a misunderstanding in how I described the problem. This code has *always* been using two contexts like I described but only in the past two days since we pulled updates to the vendor package did it start failing. There were not any changes whatsoever related to any of this code for several months besides pulling in the package update.

Only after we pulled in the update and everything started failing, we switched to contexts without timeouts and started seeing things succeed. If your explanation is correct, the defer cancel was not previously working as expected. Otherwise your explanation is not correct

**Comment by ianlancetaylor** [comment]
Do you happen to have a record of which revision of the context package you were using previously, and which one you updated to?

**Comment by bcmills** [comment]
> only in the past two days since we pulled updates to the vendor package did it start failing.

Which vendored packages did you update? (Just `x/net`, or also `x/oauth2`?)

**Comment by matthewceravolo** [comment]
I've narrowed down the precise commit where things began to fail, and there are more vendor updates here than I'd previously thought. It looks like both packages (x/net and x/oauth2) were updated and that the previous version was decently old such that it may be very difficult to narrow anything down, but here are a couple of samples from the revisions:

"checksumSHA1": "dr5+PfIRzXeN+l1VG+s0lea9qz8=",		=>	"checksumSHA1": "GtamqiJoL7PGHsN454AoffBFMa8=",
"path": "golang.org/x/net/context",	 		=>	"path": "golang.org/x/net/context",
"revision": "f5079bd7f6f74e23c4d65efa0f4ce14cbd6a3c0f",	=>	"revision": "1e491301e022f8f977054da4c2d852decd59571f",
"revisionTime": "2017-07-19T21:11:51Z"			=>	"revisionTime": "2018-05-30T06:29:46Z"

"checksumSHA1": "hyK05cmzm+vPH1OO+F1AkvES3sw=",	        => "checksumSHA1": "xaHFy3C2djwUXtiURpX3/ruY8IA=",
"path": "golang.org/x/oauth2",	 		        => "path": "golang.org/x/oauth2",
"revision": "314dd2c0bf3ebd592ec0d20847d27e79d0dbe8dd",	=> "revision": "1e0a3fa8ba9a5c9eb35c271780101fdaf1b205d7",
"revisionTime": "2016-12-14T09:25:55Z"	                => "revisionTime": "2018-06-02T17:55:33Z"

Also, I re-tested the above code by leaving everything the same and completely removing the second context.WithTimeout inside of GetCalendars such that only the client has a timeout configured, and I see the same error

**Comment by bcmills** [comment]
https://golang.org/cl/45370 is in that range and loosely matches @meirf's diagnosis. It looks like your vendor update uncovered a latent bug within your code.

**Comment by matthewceravolo** [comment]
I've confirmed that commenting out the `defer cancel()` inside of newClient does in fact successfully return our calendar data as expected, but it is still not matching what was diagnosed above since when the error occurs it is on GetCalendars, it isn't canceling prior to Calendar.New

Is the expectation that we exercise the client we retrieve to its fullest extent only to call `cancel()` at the last moment? And that if it's set on this client we shouldnt even need to bother with client.CalendarList.List().Context(), correct?

Can we rely on GC and just not bother holding onto the cancel func?

**Comment by matthewceravolo** [comment]
@fraenkel I'm fine with whichever, but doesn't putting it on the client prevent me from needing to duplicate this context code across all calls?

**Comment by bradfitz** [comment]
@matthewceravolo, please don't retitle the bug. We have a convention for how we use titles.

It still reads above like this is user error and you got lucky in the past, so I'd like to keep this bug about thinking about ways to make this debugging easier for other users making similar errors.

**Comment by Sajmani** [comment]
@bcmills I wouldn't use type assertion to introspect a Context because they are wrapped so often. If you want to attach a Done reason to a Context, I'd use a Context value.  This is overhead, though, so I'd make it opt-in in somehow. Perhaps provide a context.WithDebug method that adds a Context value for storing debugging info. The code in WithTimeout and WithCancel can check for this debug value to decide whether to walk the stack to record any information there.  The debug value needs to be installed at the top of the context stack to be seen by all derived contexts, so I'd enable it in the request handler (say) with a parameter like ?debugcontext=1.

**Comment by bcmills** [comment]
> I wouldn't use type assertion to introspect a Context because they are wrapped so often.

Good point. A `Context.Value` call with an exported (but opaque) key to retrieve the reason code should be fine.

> If you want to attach a Done reason to a Context, I'd use a Context value. This is overhead, though, so I'd make it opt-in in somehow. Perhaps provide a context.WithDebug method that adds a Context value for storing debugging info. The code in WithTimeout and WithCancel can check for this debug value to decide whether to walk the call stack and record any information there.

Ironically, `Context.Value` is O(N) with the depth of `Context` nesting, whereas `runtime.Caller(1)` is O(1): the mere act of checking whether debugging is enabled may be significantly more costly than storing the debug information unconditionally.

**Comment by Sajmani** [comment]
I would store a *debug in the context to flatten the lookup, leaving it nil
when debugging is disabled.  Associating that debug value with an
unexported context key allows the debug info to propagate through Context
implementations outside the context package.

**Comment by Xopherus** [comment]
I wonder if [errors](https://godoc.org/golang.org/x/exp/errors) might be a good solution for this problem? Rather than have duplicate information propagated in the context, just rely on the error returned when the context is cancelled to get a stack trace.

**Comment by bcmills** [comment]
@Xopherus, that's a different problem from the one here. The first error upon cancellation arises at a call to `ctx.Err()`, which is arbitrarily far away from the call to the `cancel()` function.

This issue centers on identifying the call to the `cancel` function, not the call to the `Err` method.

**Comment by Xopherus** [comment]
@bcmills I think they're actually both the same - at least the implementation would the same for `context.WithCancel`. The timeout/deadline contexts would have slightly different implementations. 

The cancelFunc returned by `context.WithCancel` is an anonymous function which sends an error to [c.cancel](https://golang.org/src/context/context.go?s=8226:8290#L232) which triggers the cancellation. That error is currently [a global var](https://golang.org/pkg/context/#pkg-variables), but what if we instead created that error within the cancelFunc using the new [experimental errors](https://godoc.org/golang.org/x/exp/errors#New) package. We'd get information about which cancelFunc was called within the error itself. I can imagine with some slight refactoring we may be able to do the same for the other contexts.

**Comment by bcmills** [comment]
@Xopherus, tons of code today uses `==` to compare against `context.Canceled`. We can't realistically change that to instantiate a new error for each call site.

**Comment by Xopherus** [comment]
Forgive my ignorance, but isn't that a big part of the errors experiment that's going on right now? I was under the impression that Go would be moving away from `==` for error inspection and to `errors.As` or `errors.Is`. I 100% agree with you that we should avoid breaking changes, so I'd like to find a way to do this in a graceful manner. Maybe that's already being discussed in other areas of the language?

**Comment by bcmills** [comment]
`errors.Is` would allow us to return per-instance errors in a new API, but we cannot assume that existing code (using the existing API) will be updated to use that check. Compatibility requires that we not break programs that rely on today's documented behavior.

**Comment by neetle** [comment]
IMHO it's a common pain point within the community to arbitrarily receive a `context.Canceled` and not know what to do with it. 
I feel like amending what `context.WithCancel` is the wrong solution though. The amount of code manually checking for cancellation is likely high. 

Would a `context.WithCanceler(context.Context) (cxt context.Context, canceler func(error))` be a decent compromise, with the argument to `canceler` propagated to all calls of `ctx.Err()`? 

There is likely to be concerns of memory leaks, but having documentation that asks users to use constants and corresponding `go vet` checks could work well.

**Comment by BentCoder** [comment]
I have same problem and don't know how to debug this. The "CURRENT" below randomly produces:

```bash
&{%!t(string=context canceled)}
*errors.errorString
```

**CURRENT**

```
// ctx is `*http.Request.Context` which come from the handler.
func Insert(ctx context.Context) error {
	ctx, cancel := context.WithTimeout(ctx, 3 * time.Millisecond)
	defer cancel()

	qry := `INSERT INTO users (name) VALUES (?)`

	res, err := r.database.ExecContext(ctx, qry, "hello")
	// ....
}
```

**WORK FINE**

```go
func Insert() error {
	ctx, cancel := context.WithTimeout(context.Background(), 3 * time.Millisecond)
	defer cancel()

	qry := `INSERT INTO users (name) VALUES (?)`

	res, err := r.database.ExecContext(ctx, qry, "hello")
	// ....
}
```

**Comment by Sajmani** [comment]
@BentCoder the difference between your CURRENT and WORKS FINE is the use of context.Background() instead of ctx. The context.Canceled error must be coming from ctx, since context.Background() is never canceled. (If the WithTimeout had expired, the error would be context.DeadlineExceeded).

**Comment by julien-may** [comment]
Are there any updates on how to continue with this topic? 

While I like the proposals from @bcmills to add a `DoneReasoner` or @neetle's one to add a description parameter to the `CancelFunc` to describe the reason for canceling a context, it only once gets beneficial when the ecosystem uses them in order to improve logging.

Adding a "stacktrace" as a context value on the other hand would already allow us to see where a new context was created and when resolving tha stacktrace context value, we could also see which one got canceled.

**Comment by nya3jp** [comment]
My project has been suffering from this issue, so I really like to see it resolved.

I think https://github.com/golang/go/issues/35791#issuecomment-649660330 by @Sajmani is a great summary of compatibility requirements. Unfortunately @neetle's proposed `context.WithCanceler` in https://github.com/golang/go/issues/26356#issuecomment-592342591 breaks API compatibility. @bcmills's proposed `DoneReasoner` in https://github.com/golang/go/issues/26356#issuecomment-404870718 keeps API compatibility, but IIUC optional interfaces are generally not liked (e.g. https://github.com/golang/go/issues/28728#issuecomment-532834952).

Just my two cents, but I have another idea. We can annotate errors if we could check if a deadline applied to a context was expired. I'm thinking of functions like this (pardon me for naive naming, it's just for explanation):

```
type CheckFunc func() bool

// WithCheckableDeadline is similar to WithDeadline, but it also returns CheckerFunc,
// which returns true if the current time is after t.
func WithCheckableDeadline(ctx Context, t time.Time) (Context, CheckFunc, CancelFunc)

// WithCheckableTimeout is similiar to WithCheckableDeadline above, but accepts a timeout.
func WithCheckableTimeout(ctx Context, d time.Duration) (Context, CheckFunc, CancelFunc)
```

We can use these functions to annotate errors on hitting timeouts, e.g.

```
func f(ctx context.Context) error {
  ctx, check, cancel := context.WithCheckableTimeout(ctx, 10 * time.Second)
  defer cancel()

  if err := someLongOperation(ctx); err != nil {
    if check() {
      return fmt.Errorf("f timeout reached: %w", err)
    }
    return err
  }
  return nil
}
```

Please note that calling `CheckFunc` is different from calling `Deadline` of an associated context and comparing it with the current time, because `Deadline` may return a deadline earlier than one applied with `context.WithDeadline` when a parent context has a deadline.

**Comment by Sajmani** [comment]
I'm going to think on this for a bit.

**Comment by bullgare** [comment]
Are there any news on it?
I see a lot of "context canceled" errors which are not handled as there is no good way to investigate the root cause for those.

**Comment by mikeschinkel** [comment]
I am running into this trying to track down why a newly generated Buffalo app gives me `context cancelled` when trying to run in terminal on macOS, but not when running within GoLand where I could more easily debug it.:-(

**Comment by bullgare** [comment]
Yes, agree.
On one of our enterprise systems we have random spikes from time to time.
And it's very hard to prove network administrators that it's network related because I'm not sure 100% about the root cause.

**Comment by jtolio** [comment]
For those of you on this thread looking for an immediate fix, using an errors.Is-compatible library that wraps errors with stack traces could be a good solution. You would need to make sure all of your code uses errors.Is and errors.As instead of error equality, and you would need to return Wrap(ctx.Err()) instead of ctx.Err() directly, but it would otherwise give you stack traces. I personally like http://pkg.go.dev/github.com/zeebo/errs/v2

**Comment by bullgare** [comment]
For me, it is potentially returned from the standard library (http).
So I am not controlling it.

**Comment by neetle** [comment]
Half the issue tends to be that we're compensating for poorly-written code here, rather than well-written code. "Just wrap the error" doesn't give an appropriate error when the context is cancelled by another thread entirely (see: spooky-action-at-a-distance)

**Comment by Sajmani** [comment]
I had less time to work on this than I hoped, sorry. Unassigning myself to open this up for others.

**Comment by paulmil1** [comment]
this is the problem i have currently and its beyond annoying,
i question myself: "Who the hell canceled my context and why?"  and it all happens in 3rd party code you cant control, good luck me at debugging this on a remote environment

**Comment by paulmil1** [comment]
imo context's Cancel() should receive argument 'reason string' so its somewhat debuggable, currenetly theres just stupid 'context canceled' constant that does not say anything about the context of cancellation

**Comment by sfllaw** [comment]
> imo context's Cancel() should receive argument 'reason string' so its somewhat debuggable, currenetly theres just stupid 'context canceled' constant that does not say anything about the context of cancellation

#46273 suggests this as context.WithCancelReason.

**Comment by Sajmani** [comment]
I prototyped an API-compatible idea to just capture the call stack of WithCancel/Timeout/Deadline and provide a way to access it as a string. Would this be enough? We'd have to measure the time & space overheads:
https://go-review.googlesource.com/c/go/+/375977

**Comment by mitar** [comment]
I would really like that one could pass optional value to `cancel` and then one can include whatever you want there (an `error`, an `error` with a stack trace, a string). So the same as how you can pass to `panic` something.

**Comment by andreimatei** [comment]
> I prototyped an API-compatible idea to just capture the call stack of WithCancel/Timeout/Deadline and provide a way to access it as a string. Would this be enough?

It's much better than nothing but, in the context of CockroachDB, it's not ideal. We too would like to pass a string as the cancellation reason. The stack trace will likely be quite ambiguous in many places, as various short-circuits will result in the same `defer cancel()` to fire.
I'm glad this issue is on the radar, though. It's a major pain for us, and a constant source of fights about the technicalities of how to hack around it.

**Comment by Sajmani** [comment]
@andreimatei @mitar I can prototype a way to pass in the string; my focus here is keeping the existing API backwards-compatible.
My interest in using the stack trace is that it improves debuggability for all _existing_ code you don't need to change any of the cancelation sites to start getting the stack traces. This comes at the expense of capturing a stack trace at each call to WithCancel/Timeout/Deadline. If that's mostly happening close to RPCs and I/O operations, this might be fine.

**Comment by andreimatei** [comment]
> My interest in using the stack trace is that it improves debuggability for all existing code you don't need to change any of the cancelation sites to start getting the stack traces. This comes at the expense of capturing a stack trace at each call to WithCancel/Timeout/Deadline. If that's mostly happening close to RPCs and I/O operations, this might be fine.

Oh, I had misunderstood what you change does. I though it captures a stack when `cancel()` is called, not when `WithCancel()` is called. The latter I think would be less useful for us than the former. I imagine the former should be doable?

> This comes at the expense of capturing a stack trace at each call to WithCancel/Timeout/Deadline.

I also hadn't fully internalized this. A call to `runtime.Callers()` would likely be too expensive for the CockroachDB code. I've just measured it on a beefy Linux machine, and for a few frames it costs 350ns per call, and the cost seems to go up steadily with the number of frames. I think this would be too much :S. We make plenty of cancelable contexts. Also I think your patch would do one extra allocation for the callers buffer, in addition to the `cancelCtx` alloc. That's be unfortunate; we're already suffering significantly because contexts basically always escape to the heap - they're our number one source of allocations, as measured by allocation source. Consider allocating the two together; something like:
```go
type cancelCtx struct {
   ...
   callers []uintptr
   callersAlloc [30]uitptr
}

func newCancelCtx(parent Context) cancelCtx {
        ctx = new(cancelCtx)
        ctx.Context = parent
	ctx.callers := ctx.callersAlloc[:]
	runtime.Callers(2, ctx.callers) // skip Callers and newCancelCtx
	return ctx
}
```

**Comment by Sajmani** [comment]
Capturing the stack trace at the cancel() call is likely to be uninteresting: usually cancel is called either via timer.AfterFunc (in the case of a deadline) or via defer, which may make it harder to debug if there are multiple defers in the same function. However, if your code often calls cancel directly from functions other than the ones that called WithCancel, I can see the use of capturing the stack trace at the cancel() call.

The same question arises for passing a string: should the string be passed to WithCancel/Deadline/Timeout, or to cancel? In either case, this would be a pretty messy API extension, so my preference would be to avoid making any API changes at all. I understand your performance concerns, though. Ideally we'd make this zero-cost by default.

**Comment by neetle** [comment]
If you're looking to pass a reason there's nothing stopping you from creating a custom context implementation. 

My core issue want with this would be to deal with poorly-written functions that take a context. Having an opt-in higher up the stack would likely be a decent trade off between performance and ease of use.

**Comment by Sajmani** [comment]
@neetle Yes, we could provide a way to enable stack-trace capture via a parent context, and propagate bit that down to child context (e.g. in propagateCancel). So this would be off by default, and users could enable it by setting a parent context with `ctx = context.EnableStackTraces(ctx)` or somesuch. This would turn on stack tracing for all child contexts, providing people with programmatic control over the whole tree (e.g. to debug specific requests), and it avoids having to add "reason" strings everywhere WithCancel or WithDeadline are called (though in theory there shouldn't be a huge number of these, either).

**Comment by jtolio** [comment]
If a "WithReason" style method is added, can I request that it take an `error` type instead of a `string`? The error could be wrapped with go1.13 errors package wrapping so what ctx.Err() returns is still errors.Is(err, context.Canceled), but that way more structured information could also be passed than just a string.

**Comment by mitar** [comment]
So we are then all arguing that #46273 should be reopen?

**Comment by Sajmani** [comment]
@mitar No, at this stage I'm just prototyping to get a feel for various ideas. My preference is to minimize API changes and improve debuggability in-place, but I understand performance concerns might make this difficult. Let's keep the discussion here for now

**Comment by andreimatei** [comment]
> Capturing the stack trace at the cancel() call is likely to be uninteresting: usually cancel is called either via `timer.AfterFunc` (in the case of a deadline) 

I don't have API specifics in mind, but I'd consider splitting the deadline case from the explicit cancellation case. When I get a `DeadlineExceeded` error, I'm mostly interested in what the respective deadline/timeout was. When an operation is explicitly canceled, I want to know who canceled it and why.

> or via defer, which may make it harder to debug if there are multiple defers in the same function.

Implicitly or explicitly, I'd try to differentiate between the `cancel` function being called after the respective operation finished (because the API says that one must always call the `cancel` function), or if the cancel is called affirmatively to cancel an ongoing operation. The former case should be as cheap as possible, because nobody is expected to get a `Canceled` error because of it (the `ctx` in question is not expected to be in use by anybody). The implementation of the latter can be more expensive, because it's expected to only happen on unhappy paths.

I know you want minimal API changes. I guess I personally would be inclined towards maximal changes :).

**Comment by andreimatei** [comment]
> If you're looking to pass a reason there's nothing stopping you from creating a custom context implementation.

We have such a context implementation, but it's not a nice one because `ctx.Err()` still returns the vanilla `context.Canceled` error (as it also does in @Sajmani 's proposed patch). It's also not nice because it forces us to allocate both our own context struct, and a `WithCancel()` context. The root of the problem for wanna-be `Context` implementers is the cancellation propagation code in the standard library, that makes it very expensive for stdlib implementations to interact with custom implementations: https://cs.opensource.google/go/go/+/refs/tags/go1.17.6:src/context/context.go;l=264
That code spawns a new goroutine to propagate cancellation from a stdlib ctx to a custom one. This is very expensive.
I've been timidly thinking of proposing to the go team to provide the required interfaces for custom implementations to participate in propagation as efficiently as the stdlib one, so that stdlib ctx can co-exist with the cowboys. I guess I'm asking now.

**Comment by Sajmani** [comment]
@andreimatei Others have suggested similar hooks, and IIRC we even prototyped some. That might be worthwhile, but IMO debugging context cancelation is a common enough problem that I'd like to make it better for everyone by default, if we can do so with acceptable cost.

**Comment by jtolio** [comment]
> I've been timidly thinking of proposing to the go team to provide the required interfaces for custom implementations to participate in propagation as efficiently as the stdlib one, so that stdlib ctx can co-exist with the cowboys. I guess I'm asking now.

Perhaps this deserves its own issue. I'd certainly be for it.

**Comment by Sajmani** [comment]
@jtolio There's prior discussion on this in #28728

**Comment by bcmills** [comment]
> Capturing the stack trace at the cancel() call is likely to be uninteresting: usually cancel is called either via timer.AfterFunc (in the case of a deadline) or via defer, which may make it harder to debug if there are multiple defers in the same function.

A given `WithContext` call site may have many possible `cancel` call sites, and while it's true that the `cancel` call is often just a deferred function, in some of the most interesting cases it is decidedly nontrivial.

For example, a server might construct a `context.WithCancel` for a background operation and store the `cancel` function in a map, and only invoke that `cancel` function when either the background operation completes or an RPC comes in to explicitly cancel it. In such a case it seems very important to distinguish between “canceled because the operation completed” and “canceled by RPC”, but the `WithCancel` call site would not distinguish them at all.

**Comment by Sajmani** [comment]
Based on the feedback from several of you, I'll prototype an alternative
that allows passing a value in via the cancel function.

**Comment by andreimatei** [comment]
> @jtolio There's prior discussion on this in #28728

I had missed the developments around #28728 and their implications. Those changes seem to let us avoid the extra goroutine, which is great! That is, as long as we don't want to override the `Done()` method. Not being able to override it means that we have to pay some extra allocations because we have to pair a stdlib `WithCancel` ctx with our implementation, but that's not the end of the world. With this in mind, I've given a renewed push to our attempts to implement richer cancellation, linked below.

> Based on the feedback from several of you, I'll prototype an alternative that allows passing a value in via the cancel function.

I'm excited to see what comes out of it!

FWIW, here's what we're trying within CockroachDB: [a colleague's PR](https://reviewable.io/reviews/cockroachdb/cockroach/66884#-Msqq5wzAz6iS_vycN9y) and [my commit](https://github.com/andreimatei/cockroach/commit/4d2a15f5a8dcf5fd099f8408dc3cee03d02fdcb8) on top of it.

**Comment by Sajmani** [comment]
I've updated my CL with a new version that prototypes adding *Cause variants of WithCancel, Deadline, and Timeout.
The purpose of this design is to provide backwards-compatibility for existing code while enabling users to add cancelation "causes" where they need them.
I welcome feedback on the API with respect to whether it addresses this issue:
https://go-review.googlesource.com/c/go/+/375977/3/src/context/context.go

I need to write more tests, particularly for cases where there are multiple causes in a context chain, and with mixed cancelation orders.
I also need to write benchmarks so we can measure slowdowns on critical paths and any new allocations.

**Comment by andreimatei** [comment]
This is wonderful!

Unless I'm missing something, your `Cause(ctx1)` implementation can return an error from `ctx2`'s cancel even if `ctx1` was not canceled when `ctx2` was canceled. In fact, it might return an error even if `ctx1.Error() == nil` (i.e. `ctx1` is not canceled). This is when a ctx on the chain from `ctx1` to `ctx2` overrode the `Done()` channel. Right? This seems surprising.
Have you considered adding protection against this in the way that `parentCancelCtx()` does?
If you don't want to do that, consider exporting `cancelCtxKey` so that 3rd party implementors can hijack `Value(cancelCtxKey)`.

**Comment by Sajmani** [comment]
@andreimatei I need to write some tests with chains of contexts and multiple causes. Hopefully that will help us see whether the returned causes are sensible or surprising. I'm explicitly propagating causes from parents to children in propagateCancel and cancelCtx.cancel—the same places where the context's error gets propagated. Glad to hear this API is directionally correct. If others agree, we'll need to take this up for proposal review.

**Comment by Sajmani** [comment]
@andreimatei I added some tests:
https://go-review.googlesource.com/c/go/+/375977/4/src/context/context_test.go#798
Do these exercise the case you were describing? (I'm not sure what ctx1 and ctx2 mean in your description).

**Comment by Sajmani** [comment]
I added test for cause vs. causeless cancels and more orderings. I also renamed things so the test is easier to follow. I think the behavior is pretty clear now: the first cancel cause sticks for all children, even if it's a nil cause.

**Comment by andreimatei** [comment]
> Do these exercise the case you were describing?

I don't think so. The cases I'm thinking of are about a custom implementation of context that overrides the `Done()` channel - e.g. an implementation that doesn't inherit the cancelation of its parent. In such cases, your `Cause` implementation can return a non-nil cause for a context who's `Err()` is `nil`. This is true for both `Cause(<custom ctx impl>)`, and for `Cause(stdlib ctx derived from custom ctx)`. I think this is surprising. In the particular case of a stdlib ctx, I think we should guarantee that, if `Err()` is `nil`, `Cause()` is also `nil`. Would you agree?

Here's a small program that demonstrates what I'm talking about.

```
package main

import (
	"context"
	"fmt"
	"time"
)

type uncanceledCtx struct {
	p context.Context
}

var _ context.Context = uncanceledCtx{}

func (u uncanceledCtx) Deadline() (deadline time.Time, ok bool) {
	return time.Time{}, false
}

func (u uncanceledCtx) Done() <-chan struct{} {
	return nil
}

func (u uncanceledCtx) Err() error {
	return nil
}

func (u uncanceledCtx) Value(key any) any {
	return u.p.Value(key)
}

func main() {
	ctx1, cancel1 := context.WithCancelCause(context.Background())
	uc := uncanceledCtx{p: ctx1}
	ctx2 := context.WithValue(uc, "key", "val")
	cancel1(fmt.Errorf("cancel 1"))

	fmt.Printf("uncanceled: Err: %v, cause: %v\n", uc.Err(), context.Cause(uc)) // uncanceled: Err: <nil>, cause: cancel 1
	fmt.Printf("ctx2: Err: %v, cause: %v\n", ctx2.Err(), context.Cause(ctx2))   // ctx2: Err: <nil>, cause: cancel 1
}
```

The output of `Cause(uc)` and `Cause(ctx2)` is surprising to me.

**Comment by Sajmani** [comment]
Ah, thanks for the reproduction case. I would state the expectation in the reverse: Cause(ctx) is only valid (defined) if ctx.Err() != nil (which is the same as when ctx.Done() is closed). If ctx.Err() is nil, then Cause(ctx) is undefined. I can add this to the spec in the CL.

We _could_ strengthen the spec for Cause to mirror what we say for Err:
```
	// If Done is not yet closed, Err returns nil.
	// If Done is closed, Err returns a non-nil error explaining why:
	// Canceled if the context was canceled
	// or DeadlineExceeded if the context's deadline passed.
	// After Err returns a non-nil error, successive calls to Err return the same error.
	Err() error
```

So we could add (and implement) "If ctx.Err returns nil, the Cause(ctx) returns nil."
I'd prefer not to add this strengthening of the spec unless there's a good reason to do so.
I expect most uses of Cause to check <-ctx.Done() or ctx.Err() != nil before calling Cause(ctx).

**Comment by andreimatei** [comment]
> I would state the expectation in the reverse: `Cause(ctx)` is only valid (defined) if `ctx.Err() != nil` (which is the same as when ctx.Done() is closed). If `ctx.Err()` is nil, then `Cause(ctx)` is undefined. I can add this to the spec in the CL.

> I expect most uses of Cause to check <-ctx.Done() or ctx.Err() != nil before calling Cause(ctx).

In my opinion, we should aim to relegate `ctx.Err()` to the past. I think we should aim for `Cause(ctx)` to completely replace it for new programs (because it's much better) (*). So I personally do not agree with the premise that any `Cause()` call should be preceded by an `Err()` call (or, rather, I don't think we should design `Cause()` with that mindset). There may be a preceding `Done()` call, but not always.

Another thing I wanted to bring up is making sure that custom ctx impls can participate in `Cause()` determination for themselves and their children. I think with your current patch that's not possible, because the key that `Cause()` looks for is not exported?

(*) If I'd have my druthers, we would make `Cause()` a new method on `ctx` (even if that means that custom Ctx impls will need to add that method to compile under the new go version. I'm not very familiar with the go compatibility promise but, as far as I'm personally concerned, there's not that many custom Ctx impls out there and the changes to them would be trivial (e.g. they can delegate to `Err()`).
I'd also consider adding a version of the `Done()` method that returns the error. It'd have to be a new channel per caller; I haven't thought through the performance implications.

**Comment by bcmills** [comment]
@andreimatei, the [Go compatibility policy](https://go.dev/doc/go1compat) does not allow new methods to be added to exported interfaces. So there is no plausible way to `Cause` to replace `Err`.

The best we could perhaps do is retrofit a top-level `func Cause(Context) error` function into the `context` package, which would use `Cause` if defined or `Err` otherwise.

**Comment by Sajmani** [comment]
Regarding compatibility, my goal with this CL is to avoid breaking anything, including existing implementations of the Context interface. Therefore I'm not adding Cause() to the Context interface.

My goal with Cause is not to replace Err, but to allow people to supplement Err with additional information.

I think you're saying that you'd like to check Cause(ctx) != nil instead of ctx.Err() != nil. If so, then indeed we'd need to make sure they are nil / not-nil at the same times. We'd probably also want to make Cause(ctx) == ctx.Err() when the user hasn't provided a cause. But this then makes it difficult to determine whether ctx.Done was closed due to cancelation or deadline exceeded (the user would need to encode this into their cause).

Regarding implementing Cause for custom contexts, users can do this easily by wrapping a context created using WithCancelCause. I'd like to keep things simple until there's a demonstrated need to support anything fancier.

**Comment by Sajmani** [comment]
All good points, @andreimatei , thank you.
I suggest at this point I write up the current state as a Proposal and document your suggestions as Alternatives to Consider.

**Comment by rsc** [comment]
Please take a look at #51365 and see if it addresses your use cases. Thanks!
